### PR TITLE
Fix the issue that gcs service client ignores error status code

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
Add status to gcs service grpc reply to avoid status information loss.
Now if grpc status is not ok, grpc client will change GrpcStatus to Ray IOError status and miss status code.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/common/status.cc [hunk 1]
```diff
@@ -13,11 +13,30 @@
...
+static const std::string kStatusCodeObjectStoreFull = "ObjectStoreFull";
+static const std::string kStatusCodeUnknownError = "Unknown error";
+static const std::string kStatusCodeNotImplemented = "NotImplemented";
+static const std::string kStatusCodeRedisError = "RedisError";
+static const std::string kStatusCodeTimedOut = "TimedOut";
+static const std::string kStatusCodeInterrupted = "Interrupted";
+static const std::string kStatusCodeUnknown = "Unknown";
+static const std::string kStatusCodeIntentionalSystemExit = "IntentionalSystemExit";
+static const std::string kStatusCodeUnexpectedSystemExit = "UnexpectedSystemExit";
+static const std::string kStatusSeparator = ": ";
```
Comment (zhijunfu) [inline src/ray/common/status.cc]: can use macro to simply code

File: src/ray/common/status.cc [hunk 2]
```diff
@@ -36,65 +55,83 @@ void Status::CopyFrom(const State *state) {
...
+      {kStatusCodeIOError, StatusCode::IOError},
+      {kStatusCodeObjectExists, StatusCode::ObjectExists},
+      {kStatusCodeObjectStoreFull, StatusCode::ObjectStoreFull},
+      {kStatusCodeUnknownError, StatusCode::UnknownError},
+      {kStatusCodeNotImplemented, StatusCode::NotImplemented},
+      {kStatusCodeRedisError, StatusCode::RedisError},
+      {kStatusCodeTimedOut, StatusCode::TimedOut},
+      {kStatusCodeInterrupted, StatusCode::Interrupted},
+      {kStatusCodeIntentionalSystemExit, StatusCode::IntentionalSystemExit},
+      {kStatusCodeUnexpectedSystemExit, StatusCode::UnexpectedSystemExit}};
```
Comment (zhijunfu) [inline src/ray/common/status.cc]: can use macro to simplify

File: src/ray/common/status.cc [hunk 3]
```diff
@@ -36,65 +55,83 @@ void Status::CopyFrom(const State *state) {
...
-    break;
+    return kStatusCodeInterrupted;
+  case StatusCode::IntentionalSystemExit:
+    return kStatusCodeIntentionalSystemExit;
+  case StatusCode::UnexpectedSystemExit:
+    return kStatusCodeUnexpectedSystemExit;
   default:
-    type = "Unknown";
-    break;
+    return kStatusCodeUnknown;
```
Comment (zhijunfu) [inline src/ray/common/status.cc]: you can define a map from code to string with macro, and then just look up the map

File: src/ray/common/status.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/common/status.cc]: fixed

File: src/ray/common/status.cc (hunk 2 above)
Comment (ffbin) [reply src/ray/common/status.cc]: fixed

File: src/ray/common/status.cc (hunk 3 above)
Comment (ffbin) [reply src/ray/common/status.cc]: fixed

**Comment by ffbin** [comment]
@pcmoritz Hi, i have modified the status.cc and add FromString function, pls check if it is convenient, thanks!  :)

File: src/ray/protobuf/gcs_service.proto [hunk 4]
```diff
@@ -23,15 +23,15 @@ message AddJobRequest {
 }
 
 message AddJobReply {
-  bool success = 1;
+  string status = 1;
```
Comment (raulchen) [inline src/ray/protobuf/gcs_service.proto]: What about just define a `Status` message in probobuf, which has 2 fields `code` and `message`?
Then we don't need that `STATUS_SEPARATOR`.

**Review by raulchen — COMMENTED** [review]
Also, can you change the PR title to `Fix the issue that gcs service client ignores error status code`, which I think is clearer.

File: src/ray/protobuf/gcs_service.proto (hunk 4 above)
Comment (ffbin) [reply src/ray/protobuf/gcs_service.proto]: fixed

**Review by raulchen — APPROVED** [review]

**Review by jovany-wang — APPROVED** [review]

**Pull request merged**